From 2bc1eb89b61786624399cf945426d1aaae9d8f73 Mon Sep 17 00:00:00 2001
From: YumeMichi <do4suki@gmail.com>
Date: Fri, 10 Feb 2017 18:30:58 +0800
Subject: [PATCH] CMParts: Add SudaMod source [2/2] * Squashed commit: commit
 b4129559ca2e27e4255124504a462ce1dd71c6f3 Author: ghbhaha <ghbhaha@gmail.com>
 Date:   Sat Dec 3 20:58:50 2016 +0800

    clear all button feature(1/2)

    Change-Id: I8cf844eca68ad9899c90feb835e07ac3c71e2873
---
 res/values-zh-rCN/sm_strings.xml     |  29 +++++++++
 res/values/sm_arrays.xml             |  39 +++++++++++++
 res/values/sm_strings.xml            |  32 ++++++++++
 res/xml/button_settings.xml          |  11 ++++
 res/xml/recents_settings.xml         |  35 +++++++++++
 src/org/sudamod/RecentsSettings.java | 110 +++++++++++++++++++++++++++++++++++
 6 files changed, 256 insertions(+)
 create mode 100644 res/values-zh-rCN/sm_strings.xml
 create mode 100644 res/values/sm_arrays.xml
 create mode 100644 res/values/sm_strings.xml
 create mode 100644 res/xml/recents_settings.xml
 create mode 100644 src/org/sudamod/RecentsSettings.java

diff --git a/res/values-zh-rCN/sm_strings.xml b/res/values-zh-rCN/sm_strings.xml
new file mode 100644
index 0000000..945b40c
--- /dev/null
+++ b/res/values-zh-rCN/sm_strings.xml
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2015 The SudaMod Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+  <string name="recents_options">最近任务</string>  
+  <string name="recents_options_title">最近任务设置</string>
+  <string name="recents_panel_settings">最近任务面板设置</string>
+  <string name="show_clear_all_recents_button_title">一键清理按钮</string>
+  <string name="show_clear_all_recents_button_summary">在最近任务面板显示一键清理按钮</string>
+  <string name="recents_clear_all_location_title">一键清理按钮位置</string>
+  <string name="recents_clear_all_location_top_right">右上方</string>
+  <string name="recents_clear_all_location_top_left">左上方</string>
+  <string name="recents_clear_all_location_top_center">上方中央</string>
+  <string name="recents_clear_all_location_bottom_right">右下方</string>
+  <string name="recents_clear_all_location_bottom_left">左下方</string>
+  <string name="recents_clear_all_location_bottom_center">下方中央</string>
+</resources>
diff --git a/res/values/sm_arrays.xml b/res/values/sm_arrays.xml
new file mode 100644
index 0000000..a921e6a
--- /dev/null
+++ b/res/values/sm_arrays.xml
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2012-2015 The MoKee OpenSource Project
+   Copyright (C) 2015 The SudaMod Project 
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+
+    <!-- clear all recents button location -->
+    <string-array name="recents_clear_all_location_entries">
+        <item>@string/recents_clear_all_location_top_right</item>
+        <item>@string/recents_clear_all_location_top_left</item>
+        <item>@string/recents_clear_all_location_top_center</item>
+        <item>@string/recents_clear_all_location_bottom_right</item>
+        <item>@string/recents_clear_all_location_bottom_left</item>
+        <item>@string/recents_clear_all_location_bottom_center</item>
+    </string-array>
+
+    <string-array name="recents_clear_all_location_values" translatable="false">
+        <item>0</item>
+        <item>1</item>
+        <item>2</item>
+        <item>3</item>
+        <item>4</item>
+        <item>5</item>
+    </string-array>
+
+</resources>
diff --git a/res/values/sm_strings.xml b/res/values/sm_strings.xml
new file mode 100644
index 0000000..03095f4
--- /dev/null
+++ b/res/values/sm_strings.xml
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2015 The SudaMod Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+
+    <!-- Recents panel settings -->
+    <string name="recents_options">Recents Options</string>    
+    <string name="recents_options_title">Recents Options Settings</string>
+    <string name="recents_panel_settings">Recents panel settings</string>
+    <string name="show_clear_all_recents_button_title">Clear All Button Tweaks</string>
+    <string name="show_clear_all_recents_button_summary">Show clear all button on recents panel</string>
+    <string name="recents_clear_all_location_title">Clear all recents location</string>
+    <string name="recents_clear_all_location_top_right">Top right</string>
+    <string name="recents_clear_all_location_top_left">Top left</string>
+    <string name="recents_clear_all_location_top_center">Top center</string>    
+    <string name="recents_clear_all_location_bottom_right">Bottom right</string>
+    <string name="recents_clear_all_location_bottom_left">Bottom left</string>
+    <string name="recents_clear_all_location_bottom_center">Bottom center</string>
+
+</resources>
diff --git a/res/xml/button_settings.xml b/res/xml/button_settings.xml
index ad28e0c..5a5620f 100644
--- a/res/xml/button_settings.xml
+++ b/res/xml/button_settings.xml
@@ -32,6 +32,17 @@
         android:persistent="false" />
 
     <PreferenceCategory
+        android:key="recents_options"
+        android:title="@string/recents_options" >
+
+    <PreferenceScreen
+        android:key="recents_panel"
+        android:title="@string/recents_options_title"
+        android:fragment="org.sudamod.RecentsSettings" />
+
+    </PreferenceCategory>
+
+    <PreferenceCategory
         android:key="navigation_bar_category"
         android:title="@string/navigation_bar_category" >
 
diff --git a/res/xml/recents_settings.xml b/res/xml/recents_settings.xml
new file mode 100644
index 0000000..c0dad2f
--- /dev/null
+++ b/res/xml/recents_settings.xml
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright (C) 2014 The CyanogenMod Project
+     Copyright (C) 2013 SlimRoms Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+
+<PreferenceScreen
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:title="@string/show_clear_all_recents_button_title">
+
+        <SwitchPreference
+            android:key="show_clear_all_recents"
+            android:title="@string/show_clear_all_recents_button_title"
+            android:summary="@string/show_clear_all_recents_button_summary"
+            android:defaultValue="true" />
+
+        <ListPreference
+            android:key="recents_clear_all_location"
+            android:title="@string/recents_clear_all_location_title"
+            android:entries="@array/recents_clear_all_location_entries"
+            android:entryValues="@array/recents_clear_all_location_values"
+            android:dependency="show_clear_all_recents" />
+
+</PreferenceScreen>
\ No newline at end of file
diff --git a/src/org/sudamod/RecentsSettings.java b/src/org/sudamod/RecentsSettings.java
new file mode 100644
index 0000000..ca749e7
--- /dev/null
+++ b/src/org/sudamod/RecentsSettings.java
@@ -0,0 +1,110 @@
+/*
+* Copyright (C) 2014 The CyanogenMod Project
+* Copyright (C) 2015 The SudaMod Project
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package org.sudamod;
+
+import android.os.Bundle;
+import android.os.UserHandle;
+import android.content.ContentResolver;
+import android.content.res.Resources;
+import android.support.v7.preference.ListPreference;
+import android.support.v7.preference.Preference;
+import android.support.v7.preference.Preference.OnPreferenceChangeListener;
+import android.support.v7.preference.PreferenceCategory;
+import android.support.v7.preference.PreferenceScreen;
+import android.support.v14.preference.SwitchPreference;
+import android.provider.Settings;
+import org.cyanogenmod.cmparts.R;
+import org.cyanogenmod.cmparts.SettingsPreferenceFragment;
+
+public class RecentsSettings extends SettingsPreferenceFragment implements OnPreferenceChangeListener {
+
+    private static final String SHOW_CLEAR_ALL_RECENTS = "show_clear_all_recents";
+    private static final String RECENTS_CLEAR_ALL_LOCATION = "recents_clear_all_location";
+
+    private SwitchPreference mRecentsClearAll;
+    private ListPreference mRecentsClearAllLocation;
+
+    @Override
+    public void onCreate(Bundle icicle) {
+        super.onCreate(icicle);
+        addPreferencesFromResource(R.xml.recents_settings);
+
+        PreferenceScreen prefSet = getPreferenceScreen();
+        ContentResolver resolver = getActivity().getContentResolver();
+
+        mRecentsClearAll = (SwitchPreference) prefSet.findPreference(SHOW_CLEAR_ALL_RECENTS);
+        mRecentsClearAll.setChecked(Settings.System.getIntForUser(resolver,
+            Settings.System.SHOW_CLEAR_ALL_RECENTS, 1, UserHandle.USER_CURRENT) == 1);
+        mRecentsClearAll.setOnPreferenceChangeListener(this);
+
+        mRecentsClearAllLocation = (ListPreference) prefSet.findPreference(RECENTS_CLEAR_ALL_LOCATION);
+        int location = Settings.System.getIntForUser(resolver,
+                Settings.System.RECENTS_CLEAR_ALL_LOCATION, 3, UserHandle.USER_CURRENT);
+        mRecentsClearAllLocation.setValue(String.valueOf(location));
+        mRecentsClearAllLocation.setOnPreferenceChangeListener(this);
+        updateRecentsLocation(location);
+       
+    }
+
+
+    public boolean onPreferenceChange(Preference preference, Object objValue) {
+        if (preference == mRecentsClearAll) {
+            boolean show = (Boolean) objValue;
+            Settings.System.putIntForUser(getActivity().getContentResolver(),
+                    Settings.System.SHOW_CLEAR_ALL_RECENTS, show ? 1 : 0, UserHandle.USER_CURRENT);
+            return true;
+        } else if (preference == mRecentsClearAllLocation) {
+            int location = Integer.valueOf((String) objValue);
+            Settings.System.putIntForUser(getActivity().getContentResolver(),
+                    Settings.System.RECENTS_CLEAR_ALL_LOCATION, location, UserHandle.USER_CURRENT);
+            updateRecentsLocation(location);
+            return true;
+        }
+        return false;
+    }
+    
+    private void updateRecentsLocation(int value) {
+        ContentResolver resolver = getContentResolver();
+        Resources res = getResources();
+        int summary = -1;
+
+        Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, value);
+
+        if (value == 0) {
+            Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, 0);
+            summary = R.string.recents_clear_all_location_top_right;
+        } else if (value == 1) {
+            Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, 1);
+            summary = R.string.recents_clear_all_location_top_left;
+        } else if (value == 2) {
+            Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, 2);
+            summary = R.string.recents_clear_all_location_top_center;            
+        } else if (value == 3) {
+            Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, 3);
+            summary = R.string.recents_clear_all_location_bottom_right;
+        } else if (value == 4) {
+            Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, 4);
+            summary = R.string.recents_clear_all_location_bottom_left;
+        } else if (value == 5) {
+            Settings.System.putInt(resolver, Settings.System.RECENTS_CLEAR_ALL_LOCATION, 5);
+            summary = R.string.recents_clear_all_location_bottom_center;            
+        }
+        if (mRecentsClearAllLocation != null && summary != -1) {
+            mRecentsClearAllLocation.setSummary(res.getString(summary));
+        }
+    }
+}
